crates/ty_ide/src/completion.rs

```diff
@@ -205,6 +234,79 @@ fn token_suffix_by_kinds<const N: usize>(
     }))
 }
 
+/// Looks for the start of a `from module import <CURSOR>` statement.
+///
+/// If found, one arbitrary token forming `module` is returned.
+fn import_from_tokens(tokens: &[Token]) -> Option<&Token> {
+    use TokenKind as TK;
+
+    /// A state used to "parse" the tokens preceding the user's cursor,
+    /// in reverse, to detect a "from import" statement.
+    enum S {
+        Start,
+        Names,
+        Module,
+    }
+
+    let mut state = S::Start;
+    let mut module_token: Option<&Token> = None;
+    // Move backward through the tokens until we get to
+    // the `from` token.
+    for token in tokens.iter().rev() {
```

Comment:
Is it possible for this function to be called for more or less arbitrary positions in the source file? Could it be triggered from a completion request at the end of a very long file? Because it looks like it might walk back arbitrarily long distances, potentially making completions slow?